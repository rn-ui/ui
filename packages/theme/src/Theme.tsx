/* tslint:disable */
import deepMerge from './tools/deepMerge';
import mergeComponentAndThemes from './tools/mergeComponentAndThemes';
import { resolveIncludes } from './tools/resolveIncludes';
import normalizeTheme from './normalizer/normalizeTheme';
import defaultTheme from './resources/default';

interface Themes {
  default: object;
  light: object;
  dark: object;
}

// @ts-ignore
const themes: Themes = {};

// Privates, ideally those should be symbols
const THEME_STYLES = '@southem.theme';

const resolveStyle = (style, baseStyle) => normalizeTheme(resolveIncludes(style, baseStyle));

/**
 * Theme
 *
 * @example
 * const defaultTheme = {};
 * const defaultThemePro = {};
 *
 * Theme.registerTheme('default', [
 *   defaultTheme,
 *   defaultThemePro,
 * ]);
 */
class Theme {

  constructor(themeStyle) {
    // @ts-ignore
    // this[THEME_STYLES] = resolveStyle(themeStyle);
  }

  /**
   * registerTheme
   * @param {String} themeName
   * @param {Array|Object} styleSheets
   */
  static registerTheme(themeName, styleSheets: Array<Object>|Object): void {
    let mergedStyleSheets = styleSheets;

    if (Array.isArray(styleSheets)) {
      mergedStyleSheets = deepMerge({}, ...styleSheets);
    }

    // themes[themeName] = new Theme(themeName, mergedStyleSheets);
    // @ts-ignore
    themes[themeName] = resolveStyle(mergedStyleSheets);
  }

  /**
   * registerDefaultTheme
   * @param {Object} styleSheets
   */
  static registerDefaultTheme(styleSheets): void {
    Theme.registerTheme('default', styleSheets);
  }

  /**
   * getTheme
   * @param {String} themeName
   * @return {Object}
   */
  static getTheme(themeName): Object {
    // @ts-ignore
    if (themeName === 'default' && !themes.default) {
      Theme.registerDefaultTheme(defaultTheme);
    }

    return themes[themeName];
  }

  /**
   * getDefaultTheme
   * @return {Object}
   */
  static getDefaultTheme(): Object {
    return Theme.getTheme('default');
  }

  /**
   * isDefine
   * @param {String} themeName
   * @return {Boolean}
   */
  static isDefine(themeName): Boolean {
    return !!themes[themeName];
  }
}

export {
  Theme as default,
  themes,
};
