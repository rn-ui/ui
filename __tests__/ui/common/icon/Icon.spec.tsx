import React from 'react';
import {
  Text,
  Image,
  ImageProps,
  TouchableOpacity,
  ImageSourcePropType,
} from 'react-native';
import {
  render,
  shallow,
  fireEvent,
  RenderAPI,
} from 'react-native-testing-library';
import Theme, {ThemeProvider} from '@southem/theme';
import {
  darkTheme,
  Icon,
  IconProps,
} from '@southem/ui';

Theme.registerDefaultTheme(darkTheme);

const Mock = (props?: IconProps): React.ReactElement<{}> => {
  return (
    <ThemeProvider
      theme={'default'}>
      <Icon {...props} />
    </ThemeProvider>
  );
};

const renderComponent = (props?: IconProps): RenderAPI => {
  return render(
    <Mock {...props} />,
  );
};

describe('@icon: matches snapshot', () => {

  describe('* interaction', () => {
    it('* stateless', () => {
      const component: RenderAPI = renderComponent({
        name: 'wifi',
      });
      const { output } = shallow(component.getByType(Icon));

      expect(output).toMatchSnapshot();
    });
  });

  describe('* appearance', () => {
    it('* empty', () => {
      const component: RenderAPI = renderComponent();
      const { output } = shallow(component.getByType(Icon));

      expect(output).toMatchSnapshot();
    });

    it('* with type', () => {
      const component: RenderAPI = renderComponent({
        name: 'access-alarm',
        type: 'octicon',
        color: 'red',
      });
      const { output } = shallow(component.getByType(Icon));

      expect(output).toMatchSnapshot();
    });

    it('* custom style (styled)', () => {
      const component: RenderAPI = renderComponent({
        size: 24,
        style: {
          fontSize: 32,
          lineHeight: 34,
        },
      });
      const { output } = shallow(component.getByType(Icon));

      expect(output).toMatchSnapshot();
    });
  });

});

/**
describe('@icon: component checks', () => {
  it('should render without issues', () => {
    const component = shallow(<Icon name='wifi'/>);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with icon type', () => {
    const component = shallow(
      <Icon
        name='alert'
        type='octicon'
        reverse
        color='red'
        iconStyle={{backgroundColor: 'peru'}}
      />,
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should have onPress event', () => {
    const onPress = jest.fn();
    const component = shallow(<Icon onPress={onPress} name='wifi'/>);
    const touchable = component.childAt(0);

    touchable.simulate('press');
    expect(onPress).toHaveBeenCalledTimes(1);
  });

  it('should apply default disabled styles', () => {
    const onPress = jest.fn();
    const component = shallow(<Icon onPress={onPress} name='wifi' disabled/>);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply custom disabled styles', () => {
    const onPress = jest.fn();
    const component = shallow(
      <Icon
        onPress={onPress}
        name='wifi'
        disabled
        disabledStyle={{backgroundColor: 'pink'}}
      />,
    );

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply container style', () => {
    const component = shallow(
      <Icon name='wifi' containerStyle={{backgroundColor: 'blue'}}/>,
    );

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply reverse styles', () => {
    const component = shallow(<Icon name='wifi' reverse/>);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should set underlayColor to color when styles when underlayColor absent', () => {
    const component = shallow(<Icon name='wifi' underlayColor={null}/>);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply raised styles', () => {
    const component = shallow(<Icon name='wifi' raised/>);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply values from theme', () => {
    const component = create(
      <ThemeProvider theme={'default'}>
        <Icon/>
      </ThemeProvider>,
    );

    expect(component.root.findByProps({testID: 'iconIcon'}).props.size).toBe(
      26,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
**/
