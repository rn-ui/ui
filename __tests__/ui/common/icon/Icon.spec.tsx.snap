// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@icon: matches snapshot * appearance * custom style (styled) 1`] = `
<ForwardRef(WrappingElement)
  color="black"
  size={24}
  style={
    Object {
      "fontSize": 32,
      "lineHeight": 34,
    }
  }
/>
`;

exports[`@icon: matches snapshot * appearance * empty 1`] = `
<ForwardRef(WrappingElement)
  color="black"
  size={24}
/>
`;

exports[`@icon: matches snapshot * appearance * with type 1`] = `
<ForwardRef(WrappingElement)
  color="red"
  name="access-alarm"
  size={24}
/>
`;

exports[`@icon: matches snapshot * interaction * stateless 1`] = `
<ForwardRef(WrappingElement)
  color="black"
  name="wifi"
  size={24}
/>
`;
